public/lesswrong-editor/src/ckeditor5-footnote/src/footnoteEditing/googleDocsFootnotesNormalizer.js

```diff
@@ -0,0 +1,261 @@
+					.map(element => writer.clone(element, true)),
+			);
+			return writer.createElement('div', { 
+				[ATTRIBUTES.footnoteItem]: '',
+				[ATTRIBUTES.footnoteId]: id,
+				[ATTRIBUTES.footnoteIndex]: index,
+			}, [newFootnoteBackLink, newFootnoteContent]);
+		});
+	}
+
+	/**
+	 * Adds `footnoteReference` `footnoteIndex`, and `footnoteId`
+	 * attributes to each footnote reference.
+	 * @param {UpcastWriter} writer
+	 * @param {Range} documentRange
+	 * @param {Object.<string, {item: Element, id: string}>} backLinks
+	 */
+	_preprocessFootnoteReferences(writer, documentRange, backLinks) {
+		const references = this._getFootnoteReferences(documentRange, backLinks);
+		for (const [ index, { item, id } ] of Object.entries(references)) {
+			writer.setAttribute(ATTRIBUTES.footnoteReference, '', item);
+			writer.setAttribute(ATTRIBUTES.footnoteIndex, index, item);
+			writer.setAttribute(ATTRIBUTES.footnoteId, id, item);
+		}
+	}
+
+	/**
+	 * Generates a new footnoteItem element for each backLink,
+	 * wraps them in a footnoteSection element then replaces the
+	 * original footnote section with the newly
+	 * created element. This clone-and-replace approach is necessary
+	 * because UpcastWriter doesn't have a method to wrap existing
+	 * content with a new element.
+	 * @param {UpcastWriter} writer
+	 * @param {Range} documentRange
+	 * @param {Object.<string, {item: Element, id: string}>} backLinks
+	 */
+	_preprocessFootnoteItems(writer, documentRange, backLinks) {
+		if(!Object.keys(backLinks).length) {
+			return;
+		}
+		if(!backLinks[1] || !backLinks[1].item) {
+			console.error("Unexpected DOM structure: missing footnote with index 1.");
+			return;
+		}
+		const firstBackLink = backLinks[1].item;
+		if(
+			!firstBackLink.parent ||
+			!firstBackLink.parent.parent ||
+			!firstBackLink.parent.parent.is('element')
+		) {
+			console.error("Unexpected DOM structure: expected footnote backLink to be nested within two parent tags.")
+			return;
+		}
+		const firstFootnote = firstBackLink.parent.parent;
+		/**
+		 * @type {number}
+		*/
+		// @ts-ignore: null case only happens for unattached elements, which we know isn't true here.
+		const firstFootnoteIndex = firstFootnote.index;
+		const { previousSibling } = firstFootnote; 
+		/**
+		 * @type {Range}
+		*/
+		// @ts-ignore DefinitelyTyped mistyping of Range
+		const footnoteSectionRange = writer.createRange(
+ 			writer.createPositionBefore(firstFootnote),
+ 			writer.createPositionAt(documentRange.end),
+		)
+		const footnoteSectionParent = firstFootnote.parent;
+		if(!footnoteSectionParent) {
+			console.error("Unexpected DOM structure: expected footnote to have a parent element.")
+			return;
+		}
+		try {
+			const footnoteItems = this._createFootnoteItems(writer, backLinks)
+			const footnoteSection = writer.createElement('div', {[ATTRIBUTES.footnoteSection]: ''}, footnoteItems);
+			for(const item of [...footnoteSectionRange.getItems()]) {
```

Comment:
You gotta tell me why the `[...foo]`.

Reply:
Whoops, no reason at all, removed 😅

Reply:
Huh, actually nevermind! Turns out using `remove` while iterating over `getItems` causes it to miss most of the items. TIL! Reverted, with a comment added.